src/main/java/dev/openfeature/sdk/ImmutableContext.java

```diff
@@ -0,0 +1,83 @@
+package dev.openfeature.sdk;
+
+import lombok.Getter;
+import lombok.ToString;
+import lombok.experimental.Delegate;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The EvaluationContext is a container for arbitrary contextual data
+ * that can be used as a basis for dynamic evaluation.
+ * The ImmutableContext is an EvaluationContext implementation which is threadsafe, and whose attributes can
+ * not be modified after instantiation.
+ */
+@ToString
+@SuppressWarnings("PMD.BeanMembersShouldSerialize")
+public final class ImmutableContext implements EvaluationContext {
+
+    @Getter
+    private final String targetingKey;
+    @Delegate
+    private final Structure structure;
+
+    /**
+     * create an immutable context with an empty targeting_key and attributes provided.
+     */
+    public ImmutableContext() {
+        this("", new HashMap<>());
+    }
+
+    /**
+     * create an immutable context with an attributes provided.
+     *
+     * @param attributes evaluation context attributes
+     */
+    public ImmutableContext(Map<String, Value> attributes) {
+        this("", attributes);
+    }
+
+    /**
+     * create an immutable context with given targetingKey and attributes provided.
+     *
+     * @param targetingKey targeting key
+     * @param attributes   evaluation context attributes
+     */
+    public ImmutableContext(String targetingKey, Map<String, Value> attributes) {
+        this.structure = new ImmutableStructure(attributes);
+        this.targetingKey = targetingKey;
+    }
+
+    @Override
+    public void setTargetingKey(String targetingKey) {
+        throw new UnsupportedOperationException("changing of targeting key is not allowed");
+    }
+
+    /**
+     * Merges this EvaluationContext objects with the second overriding the this in
+     * case of conflict.
+     *
+     * @param overridingContext overriding context
+     * @return resulting merged context
+     */
+    @Override
+    public EvaluationContext merge(EvaluationContext overridingContext) {
+        if (overridingContext == null) {
+            return new ImmutableContext(this.asMap());
+        }
+        String targetingKey = "";
+        if (this.getTargetingKey() != null && !this.getTargetingKey().trim().equals("")) {
+            targetingKey = this.getTargetingKey();
+        }
+
+        if (overridingContext.getTargetingKey() != null && !overridingContext.getTargetingKey().trim().equals("")) {
+            targetingKey = overridingContext.getTargetingKey();
+        }
+        Map<String, Value> merged = new HashMap<>();
+
+        merged.putAll(this.asMap());
+        merged.putAll(overridingContext.asMap());
```

Comment:
This merge isn't recursive (but neither is the one in the existing `MutableContext`. I wonder if we should create a separate issue to discuss that? I think I would expect it to be.

@Kavindu-Dodan @justinabrahms @beeme1mr

Reply:
If the overriding context is **immutable**, then this is handled correctly (`asMap` performs a deep copy ). But, if the overriding context is **mutable**, then we pass a map with mutable values (no deep copy). 

So yes, we need an improvement here. 

I think logic get easier if we override `clone` method from `Value.java` and replace it with `ImmutableStructure.copy` usages

Reply:
I'm not talking just about deep copying, I'm talking about recursive merge.

If there is a property on `this` called `myKey`, and it has subkeys, and there's also a property called `myKey` on `overridingContext`, then the subkeys are lost, instead of merged, I think. I think that's also true for the existing implementation, though, so it's not really an issue with this PR.

Reply:
Please file a ticket. Sounds like a bug.

Reply:
@toddbaert  can you create a follow-up issue for this ?

Reply:
Created https://github.com/open-feature/java-sdk/issues/270